lib/puppet/util.rb

```diff
@@ -25,17 +25,47 @@ module Util
 
   # Run some code with a specific environment.  Resets the environment back to
   # what it was at the end of the code.
-  def self.withenv(hash)
-    saved = ENV.to_hash
-    hash.each do |name, val|
-      ENV[name.to_s] = val
-    end
+  # Windows can store unicode chars in the environment, so need to use API calls
+  #   to get and set the environment instead of ENV due to encoding problems.
+  def self.withenv(hash, mode=:posix)
```

Comment:
Yeah, so what I'm thinking is that we have a few little helpers so that this code can be written as something like this:

```ruby
def self.withenv(hash, mode = :posix)
  saved = get_env(mode)
  set_env(hash, mode)
  yield
ensure
  set_env(saved, mode)
end
```